libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2351,6 +2351,59 @@ string YulUtilFunctions::copyArrayFromStorageToMemoryFunction(ArrayType const& _
 	});
 }
 
+string YulUtilFunctions::bytesArrayConcatFunction(TypePointers _argumentTypes)
+{
+	string functionName = "bytes_concat";
+	for (TypePointer argumentType: _argumentTypes)
+	{
+		solUnimplementedAssert(argumentType->category() == Type::Category::Array, "");
+		auto const* argumentArrayType = dynamic_cast<ArrayType const*>(argumentType);
+		solUnimplementedAssert(argumentArrayType->dataStoredIn(DataLocation::Memory), "");
+		solUnimplementedAssert(argumentArrayType->sizeOnStack() == 1, "");
+		solAssert(argumentArrayType->isByteArray(), "");
+		functionName += "_" + argumentType->identifier();
+	}
+
+	return m_functionCollector.createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(<arguments>) -> outPtr {
+				outPtr := <allocateUnbounded>()
+				let dataStart := add(outPtr, 0x20)
+				let dstMemPtr := dataStart
+				<#argument>
+				{
+					let length := <arrayLength>(<paramSrcPtr>)
+					let srcMemPtr := add(<paramSrcPtr>, 0x20)
+					for {let i := 0} lt(i, length) { i := add(i, 0x20) } {
+						let bytesUsed := sub(length, i)
+						if gt(bytesUsed, 0x20) { bytesUsed := 0x20 }
+						mstore(dstMemPtr, mload(srcMemPtr))
+						dstMemPtr := add(dstMemPtr, bytesUsed)
+						srcMemPtr := add(srcMemPtr, 0x20)
+					}
```

Comment:
Can't we use some `copyToMemory` function here?

Reply:
Actually why don't you just call `tupleEncoderPacked` for now?

Reply:
We have a `memoryCopy` helper.

Reply:
Though I think we could specialize `memoryCopy` into complete-word and arbitrary-length versions.

Reply:
I used `tupleEncoderPacked` for now
Actually, with this it might work for other types than just `bytes memory`, at least for `bytesNN` and `StringLiteral`s
Will add additional tests now